### PR TITLE
can_mcux_flexcan: update socket can instances initialization

### LINKED ISSUE #31733: Unable to build socket can with frdm_k64f
**Describe the bug**
Unable to compile socket can application with frdm_k64f board

**To Reproduce**
1. cd samples/net/sockets/can
2. west build -b frdm_k64f
3. See error

**Expected behavior**
The socket can sample should build without any error.

**Impact**
socket can sample build with errors

**Logs**
````
/home/navin/workspace/zephyrproject/zephyr/drivers/can/can_mcux_flexcan.c:815:1: warning: extra tokens at end of #include directive
  815 | #define FLEXCAN_DEVICE_SOCKET_CAN(id)     \
      | ^
In file included from ../../../../../include/sys/util_macro.h:34,
                 from ../../../../../include/sys/util.h:17,
                 from ../../../../../include/sched_priq.h:9,
                 from ../../../../../include/kernel_includes.h:23,
                 from ../../../../../include/kernel.h:17,
                 from ../../../../../include/zephyr.h:18,
                 from /home/navin/workspace/zephyrproject/zephyr/drivers/can/can_mcux_flexcan.c:9:
zephyr/include/generated/devicetree_unfixed.h:5988:57: error: expected declaration specifiers or '...' before numeric constant
 5988 | #define DT_FOREACH_OKAY_INST_nxp_kinetis_flexcan(fn) fn(0)
      |                                                         ^
../../../../../include/sys/util_internal.h:65:26: note: in definition of macro '__DEBRACKET'
   65 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
../../../../../include/sys/util_internal.h:57:2: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   57 |  __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |  ^~~~~~~~~~~~~~~~~~~~
../../../../../include/sys/util_internal.h:52:2: note: in expansion of macro '__COND_CODE'
   52 |  __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |  ^~~~~~~~~~~
../../../../../include/sys/util_macro.h:117:2: note: in expansion of macro 'Z_COND_CODE_1'
  117 |  Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |  ^~~~~~~~~~~~~
../../../../../include/devicetree.h:2074:2: note: in expansion of macro 'COND_CODE_1'
 2074 |  COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT), \
      |  ^~~~~~~~~~~
../../../../../include/sys/util_internal.h:106:36: note: in expansion of macro 'DT_FOREACH_OKAY_INST_nxp_kinetis_flexcan'
  106 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
../../../../../include/sys/util_internal.h:105:26: note: in expansion of macro 'UTIL_PRIMITIVE_CAT'
  105 | #define UTIL_CAT(a, ...) UTIL_PRIMITIVE_CAT(a, __VA_ARGS__)
      |                          ^~~~~~~~~~~~~~~~~~
../../../../../include/devicetree.h:2075:8: note: in expansion of macro 'UTIL_CAT'
 2075 |       (UTIL_CAT(DT_FOREACH_OKAY_INST_,  \
      |        ^~~~~~~~
/home/navin/workspace/zephyrproject/zephyr/drivers/can/can_mcux_flexcan.c:840:1: note: in expansion of macro 'DT_INST_FOREACH_STATUS_OKAY'
  840 | DT_INST_FOREACH_STATUS_OKAY(FLEXCAN_DEVICE_SOCKET_CAN)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/navin/workspace/zephyrproject/zephyr/drivers/can/can_mcux_flexcan.c:839:
/home/navin/workspace/zephyrproject/zephyr/drivers/can/socket_can_generic.h:120:34: warning: 'socket_can_context_1' defined but not used [-Wunused-variable]
  120 | static struct socket_can_context socket_can_context_1;
      |                                  ^~~~~~~~~~~~~~~~~~~~
/home/navin/workspace/zephyrproject/zephyr/drivers/can/socket_can_generic.h:113:26: warning: 'socket_can_api' defined but not used [-Wunused-variable]
  113 | static struct canbus_api socket_can_api = {
[130/188] Building C object zephyr/subsys/net/ip/CMakeFiles/subsys__net__ip.dir/net_context.c.obj
ninja: build stopped: subcommand failed.
````

**Environment (please complete the following information):**
  - Commit SHA: af67564573f5d7a76c703e7af57f5660c176679d

### PR DESCRIPTION
**Pull request by chen-png**
The socket can driver is initialized with only one instance
by hard code, so couldn't use DT marco to iterate all can
device instances and initialize them together.

fix:  #31733 

Signed-off-by: peng1 chen <peng1.chen@intel.com>

### CODE REVIEW

File: drivers/can/can_mcux_flexcan.c [hunk 1]
```diff
@@ -811,31 +811,32 @@ static const struct can_driver_api mcux_flexcan_driver_api = {
...
-		CANBUS_RAW_L2, NET_L2_GET_CTX_TYPE(CANBUS_RAW_L2),	\
-		CAN_MTU);						\
-
-DT_INST_FOREACH_STATUS_OKAY(FLEXCAN_DEVICE_SOCKET_CAN)
+
+#include "socket_can_generic.h"
+
+static int socket_can_init_1(const struct device *dev)
+{
+	const struct device *can_dev = DEVICE_DT_GET(DT_NODELABEL(flexcan0));
```
Comment (henrikbrixandersen) [inline drivers/can/can_mcux_flexcan.c:819]: Not that I use SocketCAN, but it's unfortunate that we bind this hard to `flexcan0`.

**Comment by alexanderwachter** [comment]
We could use the zephyr,can-primary and zephyr,can-secondary labels.

File: drivers/can/can_mcux_flexcan.c (hunk 1 above)
Comment (jenmwms) [reply drivers/can/can_mcux_flexcan.c:819]: Building on that a bit, does the use of `flexcan0` take a step backward from recent update https://github.com/zephyrproject-rtos/zephyr/commit/0519a6a0e794343b6c078d4971866fa79fa74742?

**Comment by jenmwms** [comment]
This resolves the build error for me

**Comment by chen-png** [comment]
> We could use the zephyr,can-primary and zephyr,can-secondary labels.

yes, have changed to use zephyr,can-primary label.

File: drivers/can/can_mcux_flexcan.c (hunk 1 above)
Comment (chen-png) [reply drivers/can/can_mcux_flexcan.c]: changed to use zephyr,can-primary label, could specify the detail can device by dts overlay file

**Review by LeiW000 — APPROVED** [review]

File: drivers/can/can_mcux_flexcan.c [hunk 2]
```diff
@@ -810,32 +810,33 @@ static const struct can_driver_api mcux_flexcan_driver_api = {
...
-		CAN_MTU);						\
-
-DT_INST_FOREACH_STATUS_OKAY(FLEXCAN_DEVICE_SOCKET_CAN)
+#if defined(CONFIG_NET_SOCKETS_CAN) && defined(DT_CHOSEN_ZEPHYR_CAN_PRIMARY_LABEL)
+
+#include "socket_can_generic.h"
+
+static int socket_can_init_1(const struct device *dev)
+{
+	const struct device *can_dev = device_get_binding(DT_CHOSEN_ZEPHYR_CAN_PRIMARY_LABEL);
```
Comment (henrikbrixandersen) [inline drivers/can/can_mcux_flexcan.c:819]: Please add a guard for `device_get_binding()` returning `NULL`. Otherwise looks good.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
This still feels like we're going backwards to hard code a single instance. Why can't we fix `drivers/can/socket_can_generic.h` to allow multiple instances instead?

**Review by galak — CHANGES_REQUESTED** [review]
We should add `frdm_k64f` to samples/net/sockets/can/samples.yaml

**Comment by galak** [comment]
I've done an attempt at fixing the issues in a broader sense in this PR: #31885

**Comment by galak** [comment]
Closing as this is fixed by #31885 that has been merged.

**Pull request closed without merging**